### PR TITLE
Preserve extra autocast keys on FakeTensor

### LINKED ISSUE #133924: AOTAutograd not tracing AMP correctly when using PyTorch/XLA.
### Problem

Running, for example, `torch.norm` with AMP results in a tensor with the incorrect data-type. In the program below, we expect `out` to be a `float32` tensor due to [its autocast rule][1]. However, turns out the result is actually of type `float16`.

```python
@torch.compile(backend="openxla")
def foo(x):
    with torch.cuda.amp.autocast(dtype=torch.float16):
        return torch.norm(x, p=2, dim=1)

inp = torch.rand((5, 5), dtype=torch.float16, device="xla")
out = foo(inp)
```

[1]: https://github.com/pytorch/pytorch/blob/afb3e5ed6a950fcb60b763b4eeddca9ad8d58283/aten/src/ATen/autocast_mode.h#L870-L879

Even though this looks like a PyTorch/XLA issue, inspecting the FX graph received from AOTAutograd, we can observe that AOT recorded `torch.ops.aten.norm.ScalarOpt_dim` instead of `torch.ops.aten.norm.ScalarOpt_dim_dtype`. Because of that, PyTorch/XLA has no access to the appended `at::kFloat` at the end of the call.

```python
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f16[5, 5]"):
         # File: examples/test.py:21 in foo, code: return torch.norm(x, p=2, dim=1)
        norm: "f16[5]" = torch.ops.aten.norm.ScalarOpt_dim(arg0_1, 2, [1]);  arg0_1 = None
        return (norm,)
```

### Expected Behavior

I would expect AOTAutograd to record the function that is called by the autocast layer.

### Observations

1. I can't replicate this issue with `aot_eager`. The FX graph I get from AOT is using `torch.ops.aten.linalg_vector_norm` instead of `torch.ops.aten.norm.ScalarOpt_dim`. Not sure where this happened, though.
2. A similar behavior also seems to happen with `torch.nn.Conv2d` (and, possibly, its other multi-dimensional variants), where the casting of the arguments won't get registered
3. Other operations that have autocast rules should also suffer from this, but I haven't tested them all

Because of (1), it feels like it's an XLA-specific issue inside PyTorch. I don't think it's something inside PyTorch/XLA, since this is a matter of which operation is recorded in the AOTAutograd tracing.

cc @miladm @JackCaoG @ezyang @ajam

### Versions

PyTorch version: 2.5.0a0+git7128504
Is debug build: True
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.29.2
Libc version: glibc-2.31

Python version: 3.10.14 (main, Apr 24 2024, 12:45:30) [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-6.10.4-200.fc40.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY

PyTorch/XLA version: [ac7fd44f520d586c3d95fc929134c20946e60806][2]

[2]: https://github.com/pytorch/xla/commit/ac7fd44f520d586c3d95fc929134c20946e60806

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186315

XLA:CUDA tensors can carry an extra AutocastCUDA dispatch key even though their logical device is xla. FakeTensor reconstruction currently rebuilds wrapper tensors with device_for_backend_keys=xla, which restores the default XLA key set but drops extra keys from the source tensor.

During AOT/fake tracing this means the CUDA autocast wrapper is absent. For aten.norm.ScalarOpt_dim under CUDA autocast, tracing records the original ScalarOpt_dim overload with a float16 result instead of running the autocast rule and recording ScalarOpt_dim_dtype with dtype=float32.

Thread the source dispatch key set through FakeTensor construction and cache reconstruction, but only preserve extra autocast keys beyond the logical device's default autocast key. This keeps the change scoped to tensors such as XLA:CUDA that legitimately carry an additional autocast key, without copying arbitrary source dispatch keys into fake tensors.

Fixes #133924

Generated by my agent

Test Plan:
- ninja -C build torch_python
- cmake --install build
- python -m py_compile torch/_subclasses/fake_tensor.py test/test_fake_tensor.py
- python test/test_fake_tensor.py -k test_fake_xla_cuda_autocast_dispatch_keys
- python test/test_fake_tensor.py -k fake_dispatch_keys
- python test/test_fake_tensor.py FakeTensorTest.test_fake_xla_cuda_autocast_dispatch_keys -v
- git diff --check
- lintrunner -a (failed only with pre-existing clang-tidy findings in touched C++ files)

Benchmark Results:
- Benchmarked FakeTensorMode.from_tensor(torch.empty((16, 16))) 2000 times across 7 repeats using a detached baseline worktree at HEAD and the patched tree.
- Baseline median/min: 0.2122722070/0.2083188300 s per 2000 iterations.
- Patched median/min: 0.2142668030/0.2113216890 s per 2000 iterations.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the CI failures by narrowing FakeTensor cache metadata/output propagation to only extra autocast dispatch keys, rebasing past the TensorImpl clang-tidy suppression on main, and skipping the private dispatch-key mutation cache tests under Dynamo wrapping.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed the CI guard-mismatch failures by avoiding dispatch-key reads on ordinary tensors during FakeTensor conversion, while still preserving the XLA:CUDA extra AutocastCUDA key path. I also rebuilt torch_python and reran the XLA autocast regression plus the failing Dynamo cache, export LSTM, and sparse Dynamo tests.

**Review by eellison — COMMENTED** [review]
deferring to @angelayi who is working on the XLA integration. id be great if we could get this to work with more minimal changes. You could also just implement autocast as a nested mode we insert if we're using xla tensor in tracing.

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -964,7 +1046,7 @@ def __new__(
         self.fake_mode = fake_mode
         self.constant = constant
         self.pytype = pytype
-        self.dispatch_keys = dispatch_keys
+        self.dispatch_keys = extra_dispatch_keys
```
Comment (bobrenjc93) [inline torch/_subclasses/fake_tensor.py]: This changes the meaning of `self.dispatch_keys` from the full dispatch key set passed in to only the *extra autocast* keys. That conflicts with the sole external consumer in `torch/_dynamo/guards.py:3472-3477`, which does `dispatch_keys = torch._C._dispatch_keys(value); if value.dispatch_keys is not None: dispatch_keys = value.dispatch_keys` and then uses it as the complete key set for the tensor-match guard (see `get_tensor_guard_code_part`). On the precompile path, `_unpickle_tensor` (guards.py:3938) passes the *full* serialized key set into `from_meta_and_device`, which now gets filtered down to autocast-only here. So a reconstructed fake tensor that previously carried e.g. Conjugate/Negative in `.dispatch_keys` will now report only the autocast subset (or None), and those non-autocast/non-default keys are silently dropped from the reconstructed guard — a soundness regression for that path. Please either keep the extra-keys in a separate attribute (leaving `dispatch_keys` semantics intact for guards) or update guards.py to merge rather than replace.

Comment (jansel) [reply torch/_subclasses/fake_tensor.py]: My agent says I updated the PR to keep `FakeTensor.dispatch_keys` as the full represented dispatch-key set for Dynamo/precompile guards and added a separate `FakeTensor.extra_dispatch_keys` field for wrapper-only autocast additions. The extra key now propagates through fake construction, cache reconstruction, alias/detach, symbolic fast binary outputs, and generic output wrapping. I also added regressions for full Conjugate dispatch-key preservation through `_unpickle_tensor`, XLA:CUDA AutocastCUDA propagation, alias/detach preservation, and symbolic fast add propagation.

**Comment by jansel** [comment]
My agent says updated the PR for the review feedback: `dispatch_keys` remains the full represented key set, wrapper-only autocast keys now live in `extra_dispatch_keys`, and propagation is covered through construction, cache outputs, alias/detach, symbolic fast binary outputs, and generic output wrapping. Focused fake tensor and Dynamo guard serialization regressions pass locally; `lintrunner -a` only reports the existing clang-tidy findings in touched C++ files.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I pushed an update that fixes the Dynamo-wrapped failure by skipping the intentional dispatch-key mutation test under PYTORCH_TEST_WITH_DYNAMO, centralizes extra autocast dispatch-key propagation, threads it through direct fake impl constructors, and handles PrivateUse1 autocast defaults lazily. Local focused tests passed; lintrunner -a only failed unrelated clang-tidy findings in c10/core/TensorImpl.h and torch/csrc/autograd/python_variable.cpp.

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — COMMENTED** [review]
xla tensors are deprecated (torch tpu is new approach). if this is the only justification, im not sure its worth additional complexity

**Comment by jansel** [comment]
My agent says I pushed a small follow-up for the latest review feedback: `_common_extra_autocast_dispatch_keys` now reuses the stored `FakeTensor.extra_dispatch_keys` for fake inputs on the output device, avoiding the per-op `_dispatch_keys` C call and autocast-key scan while keeping the reviewed propagation semantics. On the XLA deprecation concern, the patch remains scoped to preserving existing wrapper dispatch keys during fake tracing and keeps ordinary non-XLA tensors off the dispatch-key probing path; I am leaving the broader design call to the reviewers.

**Review by albanD — CHANGES_REQUESTED** [review]
That sounds very complex for a very narrow solution?
I assume that the exact same issue would happen with every other key on the Tensor like mathbits or tracer or conj or etc.
I would expect we would have a common solution.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated this to use a common dispatch-key preservation path instead of an AutocastCUDA-only special case. FakeTensor now keeps the represented logical dispatch key set separately from the physical extra keys that can be safely installed on the wrapper, currently Conjugate, Negative, and autocast-family keys. The physical delta is computed generically from the canonical FakeTensor key set rather than by naming one device-specific key.

I also threaded this metadata through recursive MetaConverter fakification, FX GraphPickler, Dynamo guard serialization, and the fake tensor dispatch cache so reconstructed fake tensors and cache hits keep the same represented key metadata. The recursive conversion path now uses per-tensor descriptor metadata and removes the temporary descriptor attribute after callback use, which avoids poisoning view bases and avoids retaining source tensors.

I left ZeroTensor out of the physical extra-key set after testing because installing that key on FakeTensor breaks existing fake dispatch/cache behavior; it remains represented only as logical dispatch metadata when needed. The added tests cover full Conjugate dispatch-key preservation, recursive view/base non-poisoning, XLA:CUDA autocast, PrivateUse1 autocast, GraphPickler round trip, Dynamo guard reducer behavior, cache-hit reconstruction, and MetaConverter weakref cleanup.

**Comment by jansel** [comment]
My agent says I pushed a follow-up for the latest review feedback. MetaConverter now passes descriptor metadata through an explicit optional callback kwarg instead of attaching it to temporary meta tensors, while preserving compatibility with legacy callbacks that only accept make_meta_t and device, including recursive view conversion. I also collapsed the duplicate _make_subclass call, rewrote the extra_dispatch_keys reconstruction conditional, and switched the autocast-key masks to _dispatch_key_parse so parse-only keys like AutocastXLA, AutocastMTIA, and AutocastMAIA are included. Focused FakeTensor, MetaConverter, guard serialization, graph pickler, and Dynamo-wrapped tests pass locally; lintrunner only reports the existing unrelated clang-tidy findings in c10/core/TensorImpl.h and torch/csrc/autograd/python_variable.cpp.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I pushed a follow-up for the latest review feedback. MetaConverter now computes source_desc callback support once and threads that decision through recursive fakeification, FakeTensor conversion, and GraphPickler FakeTensor restore paths, so these hot paths avoid repeated inspect.signature calls while preserving dispatch-key metadata propagation. I added focused MetaConverter and GraphPickler regressions that patch meta_utils.inspect.signature and assert the recursive path introspects once and GraphPickler restore performs zero introspections. Focused FakeTensor, GraphPickler, guard serialization, cache, py_compile, rebuild/install, and diff checks passed locally; lintrunner still only reports the existing unrelated clang-tidy findings in c10/core/TensorImpl.h and torch/csrc/autograd/python_variable.cpp.

**Comment by jansel** [comment]
@claude review these changes

File: c10/core/TensorImpl.h [hunk 2]
```diff
@@ -1480,6 +1480,14 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
    */
   void _change_backend_component_keys(c10::Device device);
 
+  /**
+   * XXX: do not use, private api!
+   * Add extra dispatch keys to this tensor.
+   */
+  void _add_extra_dispatch_keys(c10::DispatchKeySet extra_keys) {
```
Comment (albanD) [inline c10/core/TensorImpl.h:1487]: This is a big code smell. We have systems that assume you don't do this?

**Review by albanD — CHANGES_REQUESTED** [review]

File: c10/core/TensorImpl.h (hunk 2 above)
Comment (jansel) [reply c10/core/TensorImpl.h:1487]: My agent says agreed—the mutator is a code smell, and it is no longer justified by a live use case. PyTorch/XLA removed XLA:CUDA in pytorch/xla#9603, and current XLATensorImpl aborts for CUDA. I am closing this PR rather than adding a new dispatch-key mutation path.

**Comment by jansel** [comment]
My agent says this PR is obsolete and should not land. PyTorch/XLA removed XLA:CUDA in pytorch/xla#9603 as part of the 2.8 deprecation, and current XLATensorImpl explicitly aborts on CUDA. The regression now fabricates that retired configuration through the private hook added by this PR, leaving the cross-cutting dispatch-key machinery without a supported producer.
